### DISCUSSION TITLE
Package versions and repository priority

### DISCUSSION THREAD

**Discussion by proofek** [composer/composer#10339] [Q&A / Support] [answered]
Hi,

I have the following constraint defined:
```
"vendor/mypackage": "~2.394.1",
```

My understanding is it should install any versions >= 2.394.1 and < 2.395 

I have 2 repositories avail: release and feature. Feature repo has higher priority to release.

Versions 2.394.1 and 2.394.2 have been published in release repo. 
Versions dev-feature1 and dev-feature2 have been published in feature repo. 

When I am trying to install this package with the constraint as above I get the following error:
```
[WARNING]      [exec] Your requirements could not be resolved to an installable set of packages.
[WARNING]      [exec] 
[WARNING]      [exec]   Problem 1
[WARNING]      [exec]     - Root composer.json requires vendor/mypackage ~2.394.1, it is satisfiable by vendor/mypackage[2.394.2] 
from composer repo (http://myrepo/repository/composer-releases) but vendor/mypackage[dev-feature1, dev-feature2] 
from composer repo (http://myrepo/repository/composer-feature) has higher repository priority. 
The packages with higher priority do not match your constraint and are therefore not installable. 
See https://getcomposer.org/repoprio for details and assistance.
``` 

I understand that dev versions are being considered the highest version (v9999999), but if that's the case the constraint should not allow for that? What am I missing?

**Comment by stof** [comment, marked as answer]
Please read the documentation page about repo priorities (using the link given in the last line of the error message).
In particular, read the section about canonical repositories.

**Reply by proofek** [reply]
Hi,

I did, but I am confused about why would the dev branch still be considered as between the constraint versions  >= 2.394.1 and < 2.395? I would assume  v9999999 > 2.395 so it should not even consider it?

**Reply by stof** [reply]
it is not about dev branches being considered between the constraint. It is the fact that your `http://myrepo/repository/composer-feature` repository provides the package (in non-matching versions that cannot be used, but that's why it is an error) and that repository is a canonical one so lower-priority repositories are not allowed to be used for that package name.

Dev versions are not considered higher versions (there are not considered as ` v9999999` at all). The error message does not talk about higher version but about higher priority.

**Reply by stof** [reply]
@Seldaek maybe that error message should be improved to mention that the higher priority repository is canonical ?

**Reply by Seldaek** [reply]
Right, will try to tweak the wording a bit.

**Reply by proofek** [reply]
Ok, that makes a bit more sense. But does that mean if the package in question exists in the first canonical repo on the list it will just simply ignore the other repos? 

That sort of goes against what this says:
```
Now if someone publishes foo/bar 2.999 to packagist.org, suddenly Composer will pick that package as it has a higher version than your latest release (say 2.4.3), and you end up installing something you may not have meant to. If the private repository is canonical however, that 2.999 version from packagist.org will not be considered at all.
```

Unless you can only have ONE canonical repo at the time?

**Reply by stof** [reply]
You can have multiple canonical repositories at the same time, if they contain different packages.

The whole point of canonical repositories is that they forbid lower priority repos from providing packages that are in that canonical repository (that's the _only_ effect of making a repository canonical).
so indeed, _for a given package name_, you can only have one canonical repository (you could have more, but they are useless)

**Reply by stof** [reply]
And no, it does not goes against that paragraph. The first sentence describes what happens when the private repository is **not** canonical. The second sentence describes what happens for a canonical repository (remembering that the default packagist.org repository is at the lowest priority).

**Reply by proofek** [reply]
Perfect @stof ! That's now clear! Thanks a lot!

**Reply by Seldaek** [reply]
FYI, improved (hopefully) the wording in https://github.com/composer/composer/commit/4e8f3e85e2ebbb65b43606068faa53e72d61f745

**Comment by sovainfo** [comment]
By default repos are considered canonical. For later repos to be considered for packages found you need to declare the repo be non-canonical. That was the behavior in V1, for V2 you need to add "canonical': false to continue the search.

You might want to reverse the order of repos and declare your dev repo non-canonical. In addition you might have a look at stability flags.

Agree with the improvement needed for the problem description. Mentioning versions not considered first is confusing. In my opinion it should report not meeting requirements first.  It could mention that further repos are ignored due to package found in repo being declared canonical.
